File: internal/routers/topology.go
package routers

import (
	"github.com/fluent-manager/fluent-manager/internal/middleware"
	"github.com/gin-gonic/gin"
)

func registerTopologyRoutes(authed *gin.RouterGroup, h *allHandlers) {
	authed.GET("/topology/tree", middleware.RequirePermission("topology", "read"), h.Topology.GetTree)

	// Environments
	envs := authed.Group("/environments")
	{
		envs.GET("", middleware.RequirePermission("topology", "read"), h.Topology.ListEnvironments)
		envs.POST("", middleware.RequirePermission("topology", "create"), h.Topology.CreateEnvironment)
		envs.PUT("/:id", middleware.RequirePermission("topology", "update"), h.Topology.UpdateEnvironment)
		envs.DELETE("/:id", middleware.RequirePermission("topology", "delete"), h.Topology.DeleteEnvironment)
	}

	// DataCenters
	dcs := authed.Group("/datacenters")
	{
		dcs.GET("", middleware.RequirePermission("topology", "read"), h.Topology.ListDataCenters)
		dcs.GET("/:id", middleware.RequirePermission("topology", "read"), h.Topology.GetDataCenter)
		dcs.POST("", middleware.RequirePermission("topology", "create"), h.Topology.CreateDataCenter)
		dcs.PUT("/:id", middleware.RequirePermission("topology", "update"), h.Topology.UpdateDataCenter)
		dcs.DELETE("/:id", middleware.RequirePermission("topology", "delete"), h.Topology.DeleteDataCenter)
	}

	// Regions
	regions := authed.Group("/regions")
	{
		regions.GET("", middleware.RequirePermission("topology", "read"), h.Topology.ListRegions)
		regions.GET("/:id", middleware.RequirePermission("topology", "read"), h.Topology.GetRegion)
		regions.POST("", middleware.RequirePermission("topology", "create"), h.Topology.CreateRegion)
		regions.PUT("/:id", middleware.RequirePermission("topology", "update"), h.Topology.UpdateRegion)
		regions.DELETE("/:id", middleware.RequirePermission("topology", "delete"), h.Topology.DeleteRegion)
	}

	// Clusters
	clusters := authed.Group("/clusters")
	{
		clusters.GET("", middleware.RequirePermission("topology", "read"), h.Topology.ListClusters)
		clusters.GET("/:id", middleware.RequirePermission("topology", "read"), h.Topology.GetCluster)
		clusters.POST("", middleware.RequirePermission("topology", "create"), h.Topology.CreateCluster)
		clusters.PUT("/:id", middleware.RequirePermission("topology", "update"), h.Topology.UpdateCluster)
		clusters.DELETE("/:id", middleware.RequirePermission("topology", "delete"), h.Topology.DeleteCluster)
		// Match rules
		clusters.GET("/:id/rules", middleware.RequirePermission("topology", "read"), h.Topology.ListMatchRules)
		clusters.POST("/:id/rules", middleware.RequirePermission("topology", "create"), h.Topology.CreateMatchRule)
		clusters.PUT("/:id/rules/:rule_id", middleware.RequirePermission("topology", "update"), h.Topology.UpdateMatchRule)
		clusters.DELETE("/:id/rules/:rule_id", middleware.RequirePermission("topology", "delete"), h.Topology.DeleteMatchRule)
	}
}

File: web/frontend/src/api/topology.ts
import api from './client'

export const getTopologyTree = () => api.get('/topology/tree')
export const getEnvironments = () => api.get('/environments')
export const createEnvironment = (data: Record<string, any>) => api.post('/environments', data)
export const updateEnvironment = (id: number, data: Record<string, any>) => api.put(`/environments/${id}`, data)
export const deleteEnvironment = (id: number) => api.delete(`/environments/${id}`)

export const getDataCenters = () => api.get('/datacenters')
export const getDataCenter = (id: number) => api.get(`/datacenters/${id}`)
export const createDataCenter = (data: Record<string, any>) => api.post('/datacenters', data)
export const updateDataCenter = (id: number, data: Record<string, any>) => api.put(`/datacenters/${id}`, data)
export const deleteDataCenter = (id: number) => api.delete(`/datacenters/${id}`)

export const getRegions = (params?: Record<string, any>) => api.get('/regions', { params })
export const getRegion = (id: number) => api.get(`/regions/${id}`)
export const createRegion = (data: Record<string, any>) => api.post('/regions', data)
export const updateRegion = (id: number, data: Record<string, any>) => api.put(`/regions/${id}`, data)
export const deleteRegion = (id: number) => api.delete(`/regions/${id}`)

export const getClusters = (params?: Record<string, any>) => api.get('/clusters', { params })
export const getCluster = (id: number) => api.get(`/clusters/${id}`)
export const createCluster = (data: Record<string, any>) => api.post('/clusters', data)
export const updateCluster = (id: number, data: Record<string, any>) => api.put(`/clusters/${id}`, data)
export const deleteCluster = (id: number) => api.delete(`/clusters/${id}`)

export const getClusterRules = (clusterID: number) => api.get(`/clusters/${clusterID}/rules`)
export const createClusterRule = (clusterID: number, data: Record<string, any>) => api.post(`/clusters/${clusterID}/rules`, data)
export const updateClusterRule = (clusterID: number, ruleID: number, data: Record<string, any>) => api.put(`/clusters/${clusterID}/rules/${ruleID}`, data)
export const deleteClusterRule = (clusterID: number, ruleID: number) => api.delete(`/clusters/${clusterID}/rules/${ruleID}`)

export const getUserScopes = (userID: number) => api.get(`/users/${userID}/scopes`)
export const setUserScopes = (userID: number, data: Record<string, any>) => api.put(`/users/${userID}/scopes`, data)
